### ISSUE TITLE
3dNwarAdjust reverts the affine transform

### ISSUE THREAD

**Issue by salma1601** [sammba-mri/sammba-mri#107]
Because the initial warp in the nonlinear loop is the previous affine transform, when using 3dNwarpAdjust the obtained mean image inverses this affine transform. In practice using this on lemurs I end up with a template rotated in comparison to the affine template. So I think we must start the non linear loop on affine transformed files with identity warps
@nadkarni-na do you agree ?

**Comment by nadkarni-na** [comment]
Kind of.

To be clear, 3dNwarpAdjust does not go so far as to inverse the transforms; it instead calculates their overall mean, and if this is not zero (which is should be if registration calculation were ideal and unbiased, meaning not prone to systematic error), adjusts all of them to produce transforms whose overall mean is indeed zero. So in some cases there may be an inversion, though relative to each other the transforms remain the same (kind of); they are all adjusted by the same amount (kind of).

In effect, if there is ever a visible difference in brain orientation between the first non-linear mean and the affine mean, it is because the affine transforms were systematically biased (the average affine transform moved/rotated/scaled/skewed too much compared to the theoretical ideal). If this is all correct, the non-linear mean brain should have a position/orientation/size/shape that is more similar to the centre-of-mass-registered mean brain than the affine mean is. I did check a few times a while ago but it was difficult to tell given the small changes in angle and the fuzziness of the centre-of-mass-registered brain.

As for a solution...strictly speaking there isn't a need for one as the non-linear mean should be more accurate in terms of orientation etc... than the affine mean. The only irritation is that if the adjustment is too big, the dilated mask could end of being badly-positioned for the second and subsequent cycles of non-linear registration. Figuring out how to adjust it is doable but fiddly.

**Comment by salma1601** [comment]
>If this is all correct, the non-linear mean brain should have a position/orientation/size/shape that is more similar to the centre-of-mass-registered mean brain than the affine mean is.

From what I see on the lemurs template, the centre-of-mass registered mean is somehow between the nonlinear warp-adjusted mean and the affine mean. The problem is that the affine mean orientation looks similar to the bons, while the center-of-mass registered and the warp-adjusted look rotated with respect to the bons. In my understanding, the registration to the bons was done to be a in coordinate system similar to the bons, no ? so now this is lost, for instance the plane of bilateral symmetry is no more x=0

**Comment by nadkarni-na** [comment]
The registration to bons was done using the bash scripts where 3dNwarpAdjust was not included, and the option of using an external image as the first target is not in sammba (yet). In any case the shape of bons is so deformed compared to that of a real brain that any setting of a co-ordinate system similar to it will be very rough.

**Comment by salma1601** [comment]
but the bilateral plane is zero for the affine and really looks rotated after adjustment. 
affine:
![aff3_meanhead](https://user-images.githubusercontent.com/7080143/39828309-cb22ab46-53ba-11e8-9a3f-f26a57d851f6.png)
adjusted
![qw1_adjusted](https://user-images.githubusercontent.com/7080143/39828278-b445702a-53ba-11e8-98c0-3a608b06bff2.png)


I sent to you the files so you can check, it is really visible

**Comment by nadkarni-na** [comment]
Yep I saw, it's probably (very likely) because the average position of the animals relative to the acquisition volume in the scanner was not perfectly aligned sagittally. Hence an advantage to bons or any other perfectly-aligned lemur brain-like external target is to fix this.

**Comment by salma1601** [comment]
OK, so do you agree that the best version of template création pipeline would be:
 1- with an external inital target template which is not oblique 
 2- with a non linear loop starting from the affine registered heads instead of centered heads ?

**Comment by nadkarni-na** [comment]
1: yes, if available

2: no, for two reasons. i) the final result will not be an unbiased true mean. ii) this would add to reslice error unless a way could be found to directly reslice the bias-corrected images into the affine-transformed space (this isn't theoretically hard, would just involve figuring out how to convert the 3dCM output into an affine transform then concatenating it)

**Comment by salma1601** [comment]
I am not sure to follow. The lemurs template has been initialized with bons but with NwarpAdjust it reverts to the oblique orientation. So how to deal with that ?

**Comment by nadkarni-na** [comment]
The lemur template currently on NITRC used bons as the first linear registration target rather than the mean of the mass-centered brains. However, it was created using the bash scripts where 3dNwarpAdjust was not included.

What has not yet been tried is using bons as a target AND doing 3dNwarpAdjust. This could still result in a slightly oblique template, though probably less so.

If the template is still oblique, then it should just be registered to bons or rotated manually. Of course to avoid reslice error, once this final adjustment has been calculated, it should be concatenated to each of the latest transforms for each individual, and all the starting images retransformed then meaned to produce the final template.

**Comment by salma1601** [comment]
>What has not yet been tried is using bons as a target AND doing 3dNwarpAdjust. 

This is what I have done: applying nwarpadjust to the centered files from your shell pipeline
>This could still result in a slightly oblique template, though probably less so.
It is still oblique as you can see from the files I sent
>Of course to avoid reslice error, once this final adjustment has been calculated, it should be concatenated to each of the latest transforms for each individual, and all the starting images retransformed then meaned to produce the final template.

Do you mean catenating the last rigid body transform to the bons to the adjusted warps ? Makes sense, but in this case what's the need to having done à preregistration to bons at the beginning?

**Comment by salma1601** [comment]
Searching here and there I found this discussion https://github.com/poldracklab/fmriprep/issues/330.
Going into the <a href="https://afni.nimh.nih.gov/pub/dist/doc/program_help/@toMNI_Qwarpar.html" rel="nofollow">toMNI_Qwarpar</a> code I found that they used the non-linear iterations on the affine registered files to MNI, and then apply the final adjusted warps to these affine registered files. I think this is the solution we are searching for: keep the initial template location while obtaining a cohort custom template. Bellow copy pasted parts from the <a href="https://github.com/afni/afni/blob/master/src/scripts_install/%40toMNI_Qwarpar" rel="nofollow">toMNI_Qwarpar code</a> 

set basename = MNI152_1mm_uni+tlrc
set AAbasename = ${tpath}/${basename}

##### Create the list of 3dUnifize-d and affine MNI-ized datasets to proces
set dlist = ( *_uni+tlrc.HEAD )

##### A level (affine) #####
nice 3dMean -prefix Qwarp_meanA -datum float $dlist

##### B level (coarsest scale nonlinear) #####

echo 'echo $argv'                                           >> BBwarp_script
echo 'foreach lucy ( $argv )'                               >> BBwarp_script
echo '   set bbb = `basename $lucy _uni+tlrc.HEAD`'         >> BBwarp_script
echo '   nice 3dQwarp -prefix ${bbb}_uniB  -blur 0 9    \'  >> BBwarp_script
echo '                -minpatch 101            -nodset  \'  >> BBwarp_script
echo "                $AAbasename" '$lucy'                  >> BBwarp_script
echo 'end'                                                  >> BBwarp_script

tcsh ./BBwarp_script $plist &

###### adjust warps (remove mean warp) and then compute mean warped dataset
BBB_domean:
nice 3dNwarpAdjust -nwarp *_uniB_WARP+tlrc.HEAD -source $dlist -prefix Qwarp_meanB

##### C level (finer level nonlinear) #####
echo 'foreach fred ( $argv )'                                          >> CCwarp_script
echo '   set bbb = `basename $fred _uni+tlrc.HEAD`'                    >> CCwarp_script
echo '   nice 3dQwarp -prefix ${bbb}_uniC  -blur 1 6               \'  >> CCwarp_script
echo '                -inilev 2  -minpatch 49 -nodset              \'  >> CCwarp_script
echo '                -iniwarp ${bbb}_uniB_WARP+tlrc.HEAD          \'  >> CCwarp_script
echo '                Qwarp_meanB+tlrc.HEAD $fred'                     >> CCwarp_script
echo 'end'                                                             >> CCwarp_script

tcsh ./CCwarp_script $qlist[$botlist[$part]-$toplist[$part]]

CCC_domean:
nice 3dNwarpAdjust -nwarp *_uniC_WARP+tlrc.HEAD -source $dlist -prefix Qwarp_meanC

##### E level (even finer nonlinear) #####
echo 'foreach fred ( $argv )'                                            >> EEwarp_script
echo '   set bbb = `basename $fred _uni+tlrc.HEAD`'                      >> EEwarp_script
echo '   nice 3dQwarp -prefix ${bbb}_uniE  -blur 0 -2                \'  >> EEwarp_script
echo '                -inilev 7  -minpatch 13 -nodset                \'  >> EEwarp_script
echo '                -iniwarp ${bbb}_uniD_WARP+tlrc.HEAD            \'  >> EEwarp_script
echo '                Qwarp_meanD+tlrc.HEAD $fred'                       >> EEwarp_script
echo 'end'                                                               >> EEwarp_script
echo 'touch EEwarp_$argv[1]_Done'                                        >> EEwarp_script
echo 'exit 0'                                                            >> EEwarp_script

  tcsh ./EEwarp_script $qlist[$botlist[$part]-$toplist[$part]] &

EEE_domean:
nice 3dNwarpAdjust -nwarp *_uniE_WARP+tlrc.HEAD -source $dlist -prefix Qwarp_meanE

 foreach fred ( $dlist )
    set bbb = `basename $fred _uni+tlrc.HEAD`
    3dNwarpApply -nwarp ${bbb}_uniE_WARP+tlrc.HEAD \
                 -source $fred -prefix ${bbb}_uniE
 end

**Comment by salma1601** [comment]
even better: we can catenate the affine transforms to the adjusted warps (with nonlinear iterations starting from the affine registered files) and apply the catenated transfrom to the centered files to avoid reslicing. What do you think ?

**Comment by nadkarni-na** [comment]
Nope, catenating the final template --> bons transform to the final adjusted warp.

With hindsight there wasn't a need to preregister to bons. The mean of the centered brains turns out to be a good enough target.

**Comment by nadkarni-na** [comment]
The poldracklab solution seems to be a compromise where they accept an inevitable reslice error to make things easier. At the moment this would involve two reslice stages in sammba due to the centering stage passing via a reslice rather than an affine transform>

**Comment by salma1601** [comment]
>Nope, catenating the final template --> bons transform to the final adjusted warp.
With hindsight there wasn't a need to preregister to bons. The mean of the centered brains turns out to be a good enough target

I totally agree with this possible solution: no preregistration to an external template, and then a rigid body registration to fix the orientation. This solution is already do able through sammba

>The poldracklab solution seems to be a compromise where they accept an inevitable reslice error to make things easier. 

We can compute the affine transform, and the adjusted warps, and then concatenate them to avoid reslice errors no ?

**Comment by nadkarni-na** [comment]
> We can compute the affine transform, and the adjusted warps, and then concatenate them to avoid reslice errors no ?

Yes that's the ideal solution

**Comment by salma1601** [comment]
I meant the adjusted warps with nonlinear iterations starting from the affine registered files, so that the linear transform is not inverted by NwarpAdjust

**Comment by nadkarni-na** [comment]
I think the linear/affine transform **does** need to be adjusted by NwarpAdjust (via it's inclusion by concatenation in the *_WARP.nii.gz files) because 3dAllineate is also prone to systematic registration error, it's not just a problem of 3dQwarp (I think).
